File: backend/src/modules/acc/acc.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Between, Like, MoreThanOrEqual, LessThanOrEqual } from 'typeorm';
import { Acc } from './entities/acc.entity';
import { MissedCall } from './entities/missed-call.entity';
import { QueryAccDto } from './dto/query-acc.dto';
import { PaginatedResult } from '../../common/dto/pagination.dto';

@Injectable()
export class AccService {
  constructor(
    @InjectRepository(Acc)
    private accRepository: Repository<Acc>,
    @InjectRepository(MissedCall)
    private missedCallRepository: Repository<MissedCall>,
  ) {}

  /**
   * 查询 CDR 记录
   */
  async findAllAcc(queryDto: QueryAccDto): Promise<PaginatedResult<Acc>> {
    const { page, limit, srcUser, dstUser, callid, startTime, endTime, sipCode } = queryDto;
    const skip = (page - 1) * limit;

    const queryBuilder = this.accRepository.createQueryBuilder('acc');

    if (srcUser) {
      queryBuilder.andWhere('acc.src_user LIKE :srcUser', { srcUser: `%${srcUser}%` });
    }

    if (dstUser) {
      queryBuilder.andWhere('acc.dst_user LIKE :dstUser', { dstUser: `%${dstUser}%` });
    }

    if (callid) {
      queryBuilder.andWhere('acc.callid LIKE :callid', { callid: `%${callid}%` });
    }

    if (startTime) {
      queryBuilder.andWhere('acc.time >= :startTime', { startTime });
    }

    if (endTime) {
      queryBuilder.andWhere('acc.time <= :endTime', { endTime });
    }

    if (sipCode) {
      queryBuilder.andWhere('acc.sip_code = :sipCode', { sipCode });
    }

    queryBuilder.orderBy('acc.time', 'DESC');
    queryBuilder.skip(skip).take(limit);

    const [items, total] = await queryBuilder.getManyAndCount();

    return new PaginatedResult(items, total, page, limit);
  }

  /**
   * 查询未接来电
   */
  async findAllMissedCalls(queryDto: QueryAccDto): Promise<PaginatedResult<MissedCall>> {
    const { page, limit, srcUser, dstUser, startTime, endTime } = queryDto;
    const skip = (page - 1) * limit;

    const queryBuilder = this.missedCallRepository.createQueryBuilder('mc');

    if (srcUser) {
      queryBuilder.andWhere('mc.src_user LIKE :srcUser', { srcUser: `%${srcUser}%` });
    }

    if (dstUser) {
      queryBuilder.andWhere('mc.dst_user LIKE :dstUser', { dstUser: `%${dstUser}%` });
    }

    if (startTime) {
      queryBuilder.andWhere('mc.time >= :startTime', { startTime });
    }

    if (endTime) {
      queryBuilder.andWhere('mc.time <= :endTime', { endTime });
    }

    queryBuilder.orderBy('mc.time', 'DESC');
    queryBuilder.skip(skip).take(limit);

    const [items, total] = await queryBuilder.getManyAndCount();

    return new PaginatedResult(items, total, page, limit);
  }

  /**
   * 获取统计信息
   */
  async getStats(startTime?: string, endTime?: string): Promise<{
    totalCalls: number;
    missedCalls: number;
    successCalls: number;
    failedCalls: number;
    callsByHour: { hour: string; count: number }[];
    callsBySipCode: { sipCode: string; count: number }[];
  }> {
    const accQuery = this.accRepository.createQueryBuilder('acc');
    const missedQuery = this.missedCallRepository.createQueryBuilder('mc');

    if (startTime) {
      accQuery.andWhere('acc.time >= :startTime', { startTime });
      missedQuery.andWhere('mc.time >= :startTime', { startTime });
    }

    if (endTime) {
      accQuery.andWhere('acc.time <= :endTime', { endTime });
      missedQuery.andWhere('mc.time <= :endTime', { endTime });
    }

    const totalCalls = await accQuery.getCount();
    const missedCalls = await missedQuery.getCount();

    // 成功呼叫 (2xx 响应码)
    const successCalls = await this.accRepository
      .createQueryBuilder('acc')
      .where('acc.sip_code LIKE :code', { code: '2%' })
      .andWhere(startTime ? 'acc.time >= :startTime' : '1=1', { startTime })
      .andWhere(endTime ? 'acc.time <= :endTime' : '1=1', { endTime })
      .getCount();

    const failedCalls = totalCalls - successCalls;

    // 按小时统计
    const callsByHour = await this.accRepository
      .createQueryBuilder('acc')
      .select("DATE_FORMAT(acc.time, '%Y-%m-%d %H:00')", 'hour')
      .addSelect('COUNT(*)', 'count')
      .where(startTime ? 'acc.time >= :startTime' : '1=1', { startTime })
      .andWhere(endTime ? 'acc.time <= :endTime' : '1=1', { endTime })
      .groupBy('hour')
      .orderBy('hour', 'DESC')
      .limit(24)
      .getRawMany();

    // 按 SIP 响应码统计
    const callsBySipCode = await this.accRepository
      .createQueryBuilder('acc')
      .select('acc.sip_code', 'sipCode')
      .addSelect('COUNT(*)', 'count')
      .where(startTime ? 'acc.time >= :startTime' : '1=1', { startTime })
      .andWhere(endTime ? 'acc.time <= :endTime' : '1=1', { endTime })
      .groupBy('acc.sip_code')
      .orderBy('count', 'DESC')
      .getRawMany();

    return {
      totalCalls,
      missedCalls,
      successCalls,
      failedCalls,
      callsByHour,
      callsBySipCode,
    };
  }

  /**
   * 获取今日统计
   */
  async getTodayStats(): Promise<{
    totalCalls: number;
    missedCalls: number;
    successRate: number;
  }> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const todayStr = today.toISOString();

    const totalCalls = await this.accRepository
      .createQueryBuilder('acc')
      .where('acc.time >= :today', { today: todayStr })
      .getCount();

    const missedCalls = await this.missedCallRepository
      .createQueryBuilder('mc')
      .where('mc.time >= :today', { today: todayStr })
      .getCount();

    const successCalls = await this.accRepository
      .createQueryBuilder('acc')
      .where('acc.sip_code LIKE :code', { code: '2%' })
      .andWhere('acc.time >= :today', { today: todayStr })
      .getCount();

    const successRate = totalCalls > 0 ? (successCalls / totalCalls) * 100 : 0;

    return {
      totalCalls,
      missedCalls,
      successRate: Math.round(successRate * 100) / 100,
    };
  }
}

File: kam-admin-module-monitor/src/main/java/io/kamailio/admin/modules/acc/service/AccService.java
package io.kamailio.admin.modules.acc.service;

import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.modules.acc.dto.QueryAccDto;
import io.kamailio.admin.modules.acc.entity.Acc;
import io.kamailio.admin.modules.acc.entity.MissedCall;
import io.kamailio.admin.modules.acc.repository.AccRepository;
import io.kamailio.admin.modules.acc.repository.MissedCallRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import jakarta.persistence.criteria.Predicate;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Service
public class AccService {
    private final AccRepository accRepository;
    private final MissedCallRepository missedCallRepository;

    public AccService(AccRepository accRepository, MissedCallRepository missedCallRepository) {
        this.accRepository = accRepository;
        this.missedCallRepository = missedCallRepository;
    }

    public PaginatedResult<Acc> findAllAcc(QueryAccDto dto) {
        var spec = accSpec(dto);
        var page = accRepository.findAll(spec, PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), org.springframework.data.domain.Sort.by(org.springframework.data.domain.Sort.Direction.DESC, "time")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    public PaginatedResult<MissedCall> findAllMissedCalls(QueryAccDto dto) {
        var spec = missedSpec(dto);
        var page = missedCallRepository.findAll(spec, PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), org.springframework.data.domain.Sort.by(org.springframework.data.domain.Sort.Direction.DESC, "time")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    private Specification<Acc> accSpec(QueryAccDto dto) {
        return (root, q, cb) -> {
            var preds = new ArrayList<Predicate>();
            if (dto.getSrcUser() != null && !dto.getSrcUser().isBlank())
                preds.add(cb.like(root.get("srcUser"), "%" + dto.getSrcUser() + "%"));
            if (dto.getDstUser() != null && !dto.getDstUser().isBlank())
                preds.add(cb.like(root.get("dstUser"), "%" + dto.getDstUser() + "%"));
            if (dto.getCallid() != null && !dto.getCallid().isBlank())
                preds.add(cb.like(root.get("callid"), "%" + dto.getCallid() + "%"));
            if (dto.getStartTime() != null)
                preds.add(cb.greaterThanOrEqualTo(root.get("time"), Instant.parse(dto.getStartTime())));
            if (dto.getEndTime() != null)
                preds.add(cb.lessThanOrEqualTo(root.get("time"), Instant.parse(dto.getEndTime())));
            if (dto.getSipCode() != null && !dto.getSipCode().isBlank())
                preds.add(cb.equal(root.get("sipCode"), dto.getSipCode()));
            return cb.and(preds.toArray(new Predicate[0]));
        };
    }

    private Specification<MissedCall> missedSpec(QueryAccDto dto) {
        return (root, q, cb) -> {
            var preds = new ArrayList<Predicate>();
            if (dto.getSrcUser() != null && !dto.getSrcUser().isBlank())
                preds.add(cb.like(root.get("srcUser"), "%" + dto.getSrcUser() + "%"));
            if (dto.getDstUser() != null && !dto.getDstUser().isBlank())
                preds.add(cb.like(root.get("dstUser"), "%" + dto.getDstUser() + "%"));
            if (dto.getStartTime() != null)
                preds.add(cb.greaterThanOrEqualTo(root.get("time"), Instant.parse(dto.getStartTime())));
            if (dto.getEndTime() != null)
                preds.add(cb.lessThanOrEqualTo(root.get("time"), Instant.parse(dto.getEndTime())));
            return cb.and(preds.toArray(new Predicate[0]));
        };
    }

    public Map<String, Object> getStats(String startTime, String endTime) {
        var st = startTime != null ? Instant.parse(startTime) : null;
        var et = endTime != null ? Instant.parse(endTime) : null;
        long totalCalls = st != null || et != null
                ? accRepository.count((Specification<Acc>) (root, q, cb) -> {
                    var p = new ArrayList<Predicate>();
                    if (st != null) p.add(cb.greaterThanOrEqualTo(root.get("time"), st));
                    if (et != null) p.add(cb.lessThanOrEqualTo(root.get("time"), et));
                    return cb.and(p.toArray(new Predicate[0]));
                })
                : accRepository.count();
        var missedSpec = (Specification<MissedCall>) (root, q, cb) -> {
            var p = new ArrayList<Predicate>();
            if (st != null) p.add(cb.greaterThanOrEqualTo(root.get("time"), st));
            if (et != null) p.add(cb.lessThanOrEqualTo(root.get("time"), et));
            return cb.and(p.toArray(new Predicate[0]));
        };
        long missedCalls = missedCallRepository.count(missedSpec);
        long successCalls = accRepository.count((Specification<Acc>) (root, q, cb) -> {
            var list = new ArrayList<Predicate>();
            list.add(cb.like(root.get("sipCode"), "2%"));
            if (st != null) list.add(cb.greaterThanOrEqualTo(root.get("time"), st));
            if (et != null) list.add(cb.lessThanOrEqualTo(root.get("time"), et));
            return cb.and(list.toArray(new Predicate[0]));
        });
        return Map.of(
                "totalCalls", totalCalls,
                "missedCalls", missedCalls,
                "successCalls", successCalls,
                "failedCalls", totalCalls - successCalls,
                "callsByHour", List.<Map<String, Object>>of(),
                "callsBySipCode", List.<Map<String, Object>>of()
        );
    }

    public Map<String, Object> getTodayStats() {
        var today = Instant.now().truncatedTo(java.time.temporal.ChronoUnit.DAYS);
        long totalCalls = accRepository.count((Specification<Acc>) (root, q, cb) -> cb.greaterThanOrEqualTo(root.get("time"), today));
        long missedCalls = missedCallRepository.count((Specification<MissedCall>) (root, q, cb) -> cb.greaterThanOrEqualTo(root.get("time"), today));
        long successCalls = accRepository.count((Specification<Acc>) (root, q, cb) -> cb.and(cb.like(root.get("sipCode"), "2%"), cb.greaterThanOrEqualTo(root.get("time"), today)));
        double successRate = totalCalls > 0 ? Math.round((successCalls * 100.0 / totalCalls) * 100) / 100.0 : 0;
        return Map.of("totalCalls", totalCalls, "missedCalls", missedCalls, "successRate", successRate);
    }
}
